main/src/library/Iterator.flix

```diff
@@ -589,199 +517,169 @@ namespace Iterator {
     ///
-    /// Consumes the entire iterator.
+    /// Does *not* consume any elements from the iterator.
     ///
-    pub def takeWhile(f: a -> Bool \ ef, iter: Iterator[a, r]): List[a] \ { ef, Read(r) } =
-        let Iterator(done, next) = iter;
-        def loop(k) = {
-            if (done())
-                k(Nil)
-            else
-                let x = next();
-                if (f(x)) loop(ks -> k(x :: ks)) else k(Nil)
+    /// The original iterator `iter` should *not* be reused.
+    ///
+    pub def takeWhile(f: a -> Bool \ ef1, iter: Iterator[a, ef, r]): Iterator[a, ef and ef1, r]=
+        let Iterator(rc, iterF) = iter;
+        def loop(x) = match x {
+            case Ans(a) => if (f(a)) Ans(a) else Done
+            case Skip   => loop(iterF())
+            case Done   => Done
         };
-        loop(identity)
+        let iter1F = () -> loop(iterF());
+        Iterator(rc, iter1F)
 
     ///
     /// Returns `iter` without the longest prefix that satisfies the predicate `f`.
     ///
-    /// Consumes the entire iterator.
+    /// Does *not* consume any elements from the iterator.
     ///
-    pub def dropWhile(f: a -> Bool \ ef, iter: Iterator[a, r]): List[a] \ { ef, Read(r) } =
-        let Iterator(done, next) = iter;
-        match forward(x -> not f(x), done, next) {
-            case None    => Nil
-            case Some(i) => i :: toList(iter)
-        }
+    /// The original iterator `iter` should *not* be reused.
+    ///
+    pub def dropWhile(f: a -> Bool \ ef1, iter: Iterator[a, ef, r]): Iterator[a, r and ef and ef1, r] \ Write(r) =
+        let Iterator(rc, iterF) = iter;
+        let start = ref true @ rc;
+        def loop() = match iterF() {
+            case Ans(a) =>
+                if (deref start) {
+                    if (f(a)) loop() else { start:= false; Ans(a) }
+                } else {
+                    Ans(a)
+                }
+            case Skip    => loop()
+            case Done    => Done
+        };
+        let iter1F = () -> loop();
+        Iterator(rc, iter1F)
 
     ///
     /// Returns the result of applying `f` to every element in `iter` and concatenating the results.
     ///
     /// Does *not* consume any elements from the iterator.
     ///
-    @Lazy
-    pub def flatMap(f: a -> Iterator[b, r1] \ Write(r1), iter: Iterator[a, r2]): Iterator[b, r1 and r2] \ Write(r2) =
-        let Iterator(done, next) = iter;
-        let cursor = ref None @ Scoped.regionOf(iter);
-        let done1 = () -> match deref cursor {
-            case None     => match flatMapForward(f, done, next) {
-                case None     => true
-                case Some(it) =>
-                    cursor := Some(it);
-                    false
+    /// Currently `f` has to generate an iterator with region `r`.
+    ///
+    pub def flatMap(f: a -> Iterator[b, ef1, r] \ ef2, ma: Iterator[a, ef, r]): Iterator[b, ef and ef1 and ef2, r] \ Write(r) =
```

Comment:
I did not read this one yet. I remember trying for a simple implementation of `flatMap` with `append`, but I never got it to work because of the types.

Reply:
`flatMap` is still a bit of a problem. 

I've now written it using the old version of `empty` (called `polymorphiEmpty` and defined internally as a helper function), it is tough to get to compile because it stores an iterator in a ref cell. I tried to write it casting the type of the stored iterator but it caused runtime crashes in the tests.

In the longer term I think the type signature of `flatMap` may be too complicated. I'm not sure the mapped function `f` needs an "ambient" effect - i.e. `ef2` floating at the right-hand side with the slash. The type for Iterator already contains an effect `ef1` so I don't think we lose any power using that - we will need something lift further effects into an Iterator which we don't have directly at the moment (we can use `map` currently which you have suggested to help eliminate casting).

~~~

pub def flatMap(f: a -> Iterator[b, ef1, r], ma: Iterator[a, ef, r]): Iterator[b, r and ef and ef1, r] \ Write(r) =

~~~